.github/calcrom/BuildAnalyzer.cpp

```diff
@@ -0,0 +1,136 @@
+#include "BuildAnalyzer.h"
+#include "Glob.h"
+#include "ElfFile.h"
+
+string default_version("");
+
+void BuildAnalyzer::AnalyzeObject(path fname_s) {
+    string ext = fname_s.extension();
+    SourceType sourceType = ext == ".s" ? SOURCE_ASM : SOURCE_C;
+    fname_s = builddir / relative(fname_s, srcbase);
+    fname_s = fname_s.replace_extension(".o");
+    if (!exists(fname_s)) {
+        throw runtime_error("No such file: " + fname_s.string());
+    }
+
+    Elf32File elf(fname_s);
+
+    // Analyze sections
+    for (Elf32_Shdr & hdr : elf.GetSectionHeaders()) {
+        string shname = elf.GetSectionName(hdr);
+        SectionType sectionType = GetSectionType(shname);
+        if (sectionType != SECTION_OTHER) {
+            sizes[sectionType][sourceType] += (hdr.sh_size + 3) & ~3;
+            auto data = elf.ReadSectionData<unsigned>(hdr);
+#ifndef NDEBUG
+            unordered_set<unsigned> unique_addrs;
+#endif
+            for (const auto & word : data) {
+                if (word == 0) {
+                    continue; // might be a relocation
+                }
+                if (find_if(program.GetProgramHeaders().cbegin(), program.GetProgramHeaders().cend(), [&word](const auto & phdr) {
+                    return phdr.p_vaddr <= word && word < phdr.p_vaddr + phdr.p_memsz;
+                }) != program.GetProgramHeaders().cend()) {
+#ifndef NDEBUG
+                    unique_addrs.insert(word);
+#endif
+                    n_hardcoded++;
+                }
+            }
+#ifndef NDEBUG
+            if (!version.empty()) {
+                for (const auto & word : unique_addrs) {
+                    cerr << "hardcoded " << version << " pointer to " << hex << word << endl;
+                }
+            }
+#endif
+        } else if (hdr.sh_type == SHT_RELA) {
+            n_relocations += elf.GetSectionElementCount<Elf32_Rela>(hdr);
+        }
+    }
+}
+
+void BuildAnalyzer::reset() {
+    memset(sizes, 0, sizeof(sizes));
+    if (!version.empty()) {
+        sizes[SECTION_TEXT][SOURCE_ASM] = 0x800; // libsyscall.a
+    }
+    n_hardcoded = 0;
+    n_relocations = 0;
+}
+
+BuildAnalyzer::BuildAnalyzer(path &_basedir, path &_subdir, string &_version) :
+    basedir(_basedir),
+    subdir(_subdir),
+    version(_version)
+{
+    reset();
+    srcbase = basedir / subdir;
+    builddir = srcbase / "build" / version;
+    if (!exists(srcbase)) {
+        throw runtime_error("No such directory: " + srcbase.string());
+    }
+
+    string elfpat = builddir + "/*.elf";
+    Glob globber(elfpat, GLOB_TILDE | GLOB_BRACE | GLOB_NOSORT);
+    if (globber.size() == 0) {
+        throw runtime_error("unable to find an ELF file with section data");
+    }
+    program.open(globber[0], Elf32File::sections | Elf32File::programs);
+}
+
+BuildAnalyzer &BuildAnalyzer::operator()() {
+    if (analyzed) {
+        reset();
+    }
+    string pattern = srcbase.string() + "/{src,asm,lib/{src,asm},lib/{!syscall}/{src,asm}}/*.{c,s,cpp}";
```

Comment:
this needs to be updated to include nitrosdk, nns, etc.

Reply:
`{!syscall}` should cover that though